dap_aria_mapping/pipeline/data_collection/patents_citations.py

```diff
@@ -0,0 +1,268 @@
+            cited_by_results.groupby("focal_id").cited_by.apply(list).to_dict()
+        )
+
+        # don't know if this is needed but just in case - save cited by data to s3
+        if self.production:
+            logger.info("saving backward citation data to s3")
+            upload_obj(
+                self.cited_by_results_dict,
+                BUCKET_NAME,
+                "inputs/data_collection/patents/patents_cited_by_citations.json",
+            )
+
+        self.next(self.generate_forward_citation_query_chunks)
+
+    @step
+    def generate_forward_citation_query_chunks(self):
+        """Generates chunks of forward citation queries to be run in parallel"""
+        import itertools
+
+        logger.info(f"Generate {self.chunk_size} forward citation queries.")
+
+        # flatten list of 'cited by' patent ids
+        flat_cited_by_results = list(
+            set(itertools.chain(*list(self.cited_by_results_dict.values())))
+        )
+
+        # we want the backward citations of the 'cited by' patents as the forward citations of the focal patents
+        base_q = patents_backward_citation_query()
+        # generate chunk_size number of query chunks to retrieve forward citation data
+        self.query_chunks = chunk_bigquery_query(
+            publication_numbers=flat_cited_by_results,
+            chunk_size=self.chunk_size,
+            base_q=base_q,
+            production=self.production,
+        )
+
+        self.next(self.retrieve_forward_citation_data, foreach="query_chunks")
+
+    @step
+    def retrieve_forward_citation_data(self):
+
+        logger.info(f"Retrieving forward citation data for focal ids.")
+        # query bigquery to retrieve forward citation data in chunks
+        self.forward_citation_chunks = conn.query(self.input).to_dataframe()
+
+        self.next(self.join)
+
+    @step
+    def join(self, inputs):
+        """Join and clean forward citation results"""
+
+        # join forward citation results
+        logger.info(f"join {self.chunk_size} forward citation results")
+        forward_citation_results = pd.concat(
+            [i.forward_citation_chunks for i in inputs]
+        )
+
+        logger.info(f"{forward_citation_results.shape}")
+
+        # generate forward citations dictionary
+        self.forward_citations = (
+            forward_citation_results.groupby("focal_id")
+            .patent_citation.apply(list)
+            .to_dict()
+        )
+
+        self.next(self.end)
+
+    @step
+    def end(self):
+        """saves forward citations to s3"""
+        from nesta_ds_utils.loading_saving.S3 import upload_obj
+        from dap_aria_mapping import BUCKET_NAME
+
+        # save forward citations to s3
+        if self.production:
+            logger.info("saving forward citation data to s3")
+            upload_obj(
+                self.forward_citations,
+                BUCKET_NAME,
+                "inputs/data_collection/patents/patents_forward_citations.json",
```

Comment:
Maybe save different years as separate files and include the year in the file name